doc/configuration.md
# Configuration File

This is the default configuation file.

    # === [ MySQL Settings ] ===
    MySQL:
      # activate MySQL
      active: false
      # hostname to use to connect to the database, in most cases 'localhost'
      hostname: host
      # username to use to connect to the database
      username: user
      # password to use to connect to the database
      password: pw
      # database name to connect to
      database: db
      # database port to connect to
      port: 3306
      # general statistic table name
      table: pvpstats
      # kill statistic table name
      killtable: pvpkillstats
      # connection options
      options: ?autoReconnect=true
    # === [ SQLite Settings ] ===
    SQLite:
      # activate SQLite
      active: false
      # general statistic table name
      table: pvpstats
      # kill statistic table name
      killtable: pvpkillstats
    # === [ YML Database Emulation Settings ] ===
    YML:
      # general statistic table name
      table: pvpstats
      # kill statistic table name
      killtable: pvpkillstats
    # === [ Statistic Settings ] ===
    statistics:
      # clear (duplicated) statistics on every start
      clearOnStart: true
      # prevent players from getting kills from the same victim
      checkAbuse: true
      # seconds to wait before allowing to kill the same player again to count (-1 will never reset)
      abuseSeconds: -1
      # save every kill - is never read internally, so only for web stats or alike
      collectPrecise: true
      # always reset a streak when a player disconnects
      resetKillstreakOnQuit: false
      # count dying from other sources than players towards death count and resetting of streaks
      countRegularDeaths: false
      # mathematical formula to calculate kill/death ratio
      killDeathCalculation: '&k/(&d+1)'
    # === [ Integration into other Plugins ] ===
    other:
      # count PVP Arena deaths
      PVPArena: false
    # world names where not to count statistics
    ignoreworlds:
      - doNotTrack
    # === [ ELO Score Settings ] ===
    eloscore:
      active: false
      # min possible ELO score
      minimum: 18
      # starting ELO score
      default: 1500
      # max possible ELO score
      maximum: 3000
      # K-Factor settings
      k-factor:
        # K-Factor below threshold
        below: 32
        # K-Factor above threshold
        above: 16
        # K-Factor threshold
        threshold: 2000
    # === [ Message Settings ] ===
    msg:
      overrides: false
      main:
        - '&cName: &7%n'
        - '&cKills: &7%k'
        - '&cDeaths: &7%d'
        - '&cRatio: &7%r'
        - '&cStreak: &7%s'
        - '&cMax Streak: &7%m'
    # locations of leaderboards
    leaderboards: []
    # === [ Updater Settings ] ===
    update:
      # phone home to inform slipcor that you use the plugin
      tracker: true
      # what to do? Valid values: disable, announce, download, both
      mode: both
      # which type of branch to get updates? Valid values: dev, alpha, beta, release
      type: beta


doc/commands/config.md
# Config Command

## Description

This command allows to show and edit config settings

## Usage Examples

Command |  Definition
------------- | -------------
/pvpstats config get [node] | get the value of a config node
/pvpstats config set [node] [value] | set the value of a config node
/pvpstats config add [node] [value] | add an entry to a config list
/pvpstats config remove [node] [value] | remove an entry from a config list

## Details

The plugin will do its best to auto tab complete the node, always make sure you use tab or know what you are doing.



doc/changelog.md
# v1.X Changelog
- v1.11.12 - remove deprecated SQL driver, this loading is no longer necessary anyway
- v1.11.11 - improve update config setting comment, fix all comments having duplicated comment characters
- v1.11.10 - only clear caches instead of pre-caching all players who might have stats
- v1.11.9 - match player names more leniently
- v1.11.8 - try a more sturdy and database based approach to find OfflinePlayers, for bigger servers
- v1.11.7 - maybe prevent OfflinePlayer lookups crashing servers
- v1.11.6 - support null OfflinePlayers, because that is a thing
- v1.11.5 - update plugin.yml to reflect version bump
- v1.11.4 - properly create the language file for new users****
- v1.11.3 - remove debug code
- v1.11.2 - activate encryption!
- v1.11.1 - two updates:
  - **CHANGE**: Move around language nodes
  - **FIX**: (Core) do not inform people when they use an experimental version
- v1.10.4 - fix language file default values - sorry to new users!
- v1.10.3 - fix config command tab completion
- v1.10.2 - improve world column check and support late loading of worlds, this should no longer throw around warnings and errors
- v1.10.1 - refactor several features into a library that gets shaded in
- v1.9.6 - fix issue with mysql table checks with limited permissions
- v1.9.5 - default the "show" permission to true
- v1.9.4 several updates:
  - **FIX**: Replace hardcoded messages with language messages
  - **FEAT**: Introduce pvpstats.show to restrict players watching other players' stats
  - **FEAT**: Add new config setting 'shortHandCommands' to disable exclamation point shorthand functionality
- v1.9.3 - take minimum and default ELO into account when creating player statistics; fix test logic
- v1.9.2 - teach K/D interpreter to display a number that makes some sense rather than 0, when dividing by 0
- v1.9.1 - introduce K/D interpreter - it supports powers `x² => (x^2)` - thanks to garbagemule for the inspiration
- v1.8.18 - properly treat the "simple calculation" config node as boolean
- v1.8.17 - use plugin prefix in more places, this should reduce confusion and clarify who is talking
- v1.8.16 - hide ELO information in outputs if ELO is not activated
- v1.8.15 - allow to count pet kills, fix newbie check for deaths not from PVP
- v1.8.14 - add ability to run multiple commands on a streak
- v1.8.13 - run streak commands synchronously!
- v1.8.12 - shift the logic around to make meta restrictions work
- v1.8.11 - add configurability for numbered entries, support player meta tags (NPC) to exclude from statistics
- v1.8.10 - update version check to work with alpha testing builds
- v1.8.9 - add debug logs for async or sync
- v1.8.8 - hotfix: fix crashes for mysql users
- v1.8.7 - change default behavior for precise stats to not run. there is an override config setting that will be announced instead of a warning that no-one listens to
- v1.8.6 - (ci) fixed discord notification for multiline commit messages
- v1.8.5 - (ci) separate master and pull request ci jobs
- v1.8.4 - fix server crash when looking up player info on async compatible systems
- v1.8.3 - bump minor version for the new release system. Thanks to Ffaen for all the help :)
- v1.7.42 - Add github actions file and change pvparena repo url
- v1.7.41 - remove debug outputs to console
- v1.7.40 - reintroduce PAPI hook
- v1.7.39 - rename API method, add API documentation
- v1.7.38 - add an event to hook into and cancel 
- v1.7.37 - buffer top and flop lists
- v1.7.36 - fix mysql error
- v1.7.35 - Reintroduce **/pvpstats cleanup** to remove duplicates
- v1.7.34 - Create SQL tables properly with new victim logic
- v1.7.33 - various updates as follows
  - **FIX**: debugger not closing debug files properly
  - **FIX**: NPEs related to no-pvp-deaths and more
  - **FIX**: non-unique config nodes being offered as autocomplete and config set argument
  - **FIX**: delay kill streak and ELO messages until after the death message has been sent
  - **FEAT**: player display name feature - uses player names formatted by other plugins
  - **FEAT**: revamp autocomplete system to allow for partial matches, not only beginning-word-matches
  - **FEAT**: TextFormatter API that allows for more distinct highlights and information messages, allow information in hover-over, and issuing commands on clicks 
- v1.6.32 - Make some entries write-only for safety reasons (only mysql connection data for now)
- v1.6.31 - Rework config command, adjust op messages to provide more information, fix NPE for empty player entry
- v1.6.30 - Update debugkill, migrate DatabaseAPI to use OfflinePlayer instead of Player
- v1.6.29 - add "configset" command to set a config setting
- v1.6.28 - move usage tracking to bStats, defaulting to give some more information about which features are being used
- v1.6.27 - fix /pvpstats top (showing K/D)
- v1.6.26 - properly implement newbie check - I swear this time it works like intended
- v1.6.25 - prepare plugin.yml for proper warnings in the future
- v1.6.24 - hack-fix too greedy permissions
- v1.6.23 - fix npe on first launch
- v1.6.21 - vanished check actually reversing the logic
- v1.6.20 - hide vanished players in command tab completion
- v1.6.19 - reduce database connections in generally, especially by caching UUIDs and legacy player names
- v1.6.18 - add "killDeathSimple" config node to use simple kills/deaths instead of a complicated mathematical function
- v1.6.17 - add "deathsDescending" config node to allow for default mathematical sorting of deaths, round ratios in all places!
- v1.6.16 - only write asynchronously to mySQL
- v1.6.15 - change the newbie permission to "pvpstats.nonewbie" to actually work like a permission, defaulting to "true"
- v1.6.14 - add async runnables
- v1.6.13 - move some database interaction to async thread
- v1.6.12 - add config setting to enable the newbie feature. for backwards compatibility this will stay "true" for a while

src/main/java/net/slipcor/pvpstats/yml/Config.java
package net.slipcor.pvpstats.yml;

import net.slipcor.core.ConfigEntry;
import net.slipcor.core.CoreConfig;
import net.slipcor.core.CorePlugin;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

/**
 * Config access class
 *
 * @author slipcor
 */
public class Config extends CoreConfig {

    /**
     * The Config entry class
     *
     * Each entry has an explicit class type, a node, optional comment
     * and a default value
     */
    public enum Entry implements ConfigEntry {

        GENERAL(Type.COMMENT, "General", null,
                "=== [ General Settings ] ==="),
        GENERAL_SHORTHAND_COMMANDS(Type.BOOLEAN, "general.shortHandCommands", true,
                "Allow command shorthands"),

        MYSQL(Type.COMMENT, "MySQL", null,
                "=== [ MySQL Settings ] ==="),
        MYSQL_ACTIVE(Type.BOOLEAN, "MySQL.active", false,
                "activate MySQL"),
        MYSQL_HOST(Type.STRING, "MySQL.hostname", "host",
                "hostname to use to connect to the database, in most cases 'localhost'"),
        MYSQL_USERNAME(Type.STRING, "MySQL.username", "user",
                "username to use to connect to the database"),
        MYSQL_PASSWORD(Type.STRING, "MySQL.password", "pw",
                "password to use to connect to the database"),
        MYSQL_DATABASE(Type.STRING, "MySQL.database", "db",
                "database name to connect to"),
        MYSQL_PORT(Type.INT, "MySQL.port", 3306,
                "database port to connect to"),
        MYSQL_TABLE(Type.STRING, "MySQL.table", "pvpstats",
                "general statistic table name"),
        MYSQL_KILLTABLE(Type.STRING, "MySQL.killtable", "pvpkillstats",
                "kill statistic table name"),
        MYSQL_OPTIONS(Type.STRING, "MySQL.options", "?autoReconnect=true",
                "connection options"),

        SQLITE(Type.COMMENT, "SQLite", null,
                "=== [ SQLite Settings ] ==="),
        SQLITE_ACTIVE(Type.BOOLEAN, "SQLite.active", false,
                "activate SQLite"),
        SQLITE_FILENAME(Type.STRING, "SQLite.filename", "database",
                "database file name, plugin will append extension '.db'"),
        SQLITE_TABLE(Type.STRING, "SQLite.table", "pvpstats",
                "general statistic table name"),
        SQLITE_KILLTABLE(Type.STRING, "SQLite.killtable", "pvpkillstats",
                "kill statistic table name"),

        YML(Type.COMMENT, "YML", null,
                "=== [ YML Database Emulation Settings ] ==="),
        YML_TABLE(Type.STRING, "YML.table", "pvpstats",
                "general statistic file name, plugin will append extension '.yml'"),
        YML_KILLTABLE(Type.STRING, "YML.killtable", "pvpkillstats",
                "kill statistic file name, plugin will append extension '.yml'"),
        YML_COLLECT_PRECISE(Type.BOOLEAN, "YML.collectPreciseOverride", false,
                "really save every kill? This is not advised for YML!"),

        STATISTICS(Type.COMMENT, "statistics", null,
                "=== [ Statistic Settings ] ==="),
        STATISTICS_ASSIST_SECONDS(Type.INT, "statistics.assistSeconds", 60,
                "time in seconds to consider a former kill an assist"),
        STATISTICS_CLEAR_ON_START(Type.BOOLEAN, "statistics.clearOnStart", true,
                "clear (duplicated) statistics on every start"),
        STATISTICS_CREATE_ON_JOIN(Type.BOOLEAN, "statistics.createOnJoin", true,
                "create empty player entry when they join"),
        STATISTICS_COLLECT_PRECISE(Type.BOOLEAN, "statistics.collectPrecise", true,
                "save every kill - is never read internally, so only for web stats or alike"),
        STATISTICS_COUNT_REGULAR_DEATHS(Type.BOOLEAN, "statistics.countRegularDeaths", false,
                "count dying from other sources than players (and their tamed pets) towards death count and resetting of streaks"),
        STATISTICS_COUNT_PET_DEATHS(Type.BOOLEAN, "statistics.countPetDeaths", false,
                "count dying from other player's tamed pets for death count and resetting of streaks"),
        STATISTICS_CHECK_ABUSE(Type.BOOLEAN, "statistics.checkAbuse", true,
                "prevent players from getting kills from the same victim"),
        STATISTICS_CHECK_NEWBIES(Type.BOOLEAN, "statistics.checkNewbies", true,
                "prevent stats for players with the permission 'pvpstats.newbie'"),
        STATISTICS_ABUSE_SECONDS(Type.INT, "statistics.abuseSeconds", -1,
                "seconds to wait before allowing to kill the same player again to count (-1 will never reset)"),
        STATISTICS_DEATHS_DESCENDING(Type.BOOLEAN, "statistics.deathsDescending", false,
                "order deaths descending by default (rather than TOP deaths being LESS deaths)"),
        STATISTICS_KD_CALCULATION(Type.STRING, "statistics.killDeathCalculation", "&k/&d",
                "mathematical formula to calculate kill/death ratio"),
        STATISTICS_KD_SIMPLE(Type.BOOLEAN, "statistics.killDeathSimple", false,
                "simplify kill/death ratio calculation to just kills per deaths"),
        STATISTICS_PREVENTING_PLAYER_META(Type.LIST, "statistics.preventingPlayerMeta", Collections.singletonList("NPC"),
                "player meta that marks players as not counting, for example 'NPC' for most plugins that add them"),
        STATISTICS_RESET_KILLSTREAK_ON_QUIT(Type.BOOLEAN, "statistics.resetKillstreakOnQuit", false,
                "always reset a streak when a player disconnects"),
        STATISTICS_LEADERBOARD_REFRESH(Type.INT, "statistics.leaderboardRefresh", 60,
                "seconds to wait before the top/flop leaderboard values will query the database again"),
        STATISTICS_STREAK_ANNOUNCEMENTS(Type.BOOLEAN, "statistics.streakAnnouncements", false,
                "announce lines from streak_announcements.yml on specific streak values"),
        STATISTICS_STREAK_COMMANDS(Type.BOOLEAN, "statistics.streakCommands", false,
                "issue commands from streak_commands.yml on specific streak values"),


        ELO(Type.COMMENT, "eloscore", null,
                "=== [ ELO Score Settings ] ==="),
        ELO_ACTIVE(Type.BOOLEAN, "eloscore.active", false, null),
        ELO_MINIMUM(Type.INT, "eloscore.minimum", 18,
                "min possible ELO score"),
        ELO_DEFAULT(Type.INT, "eloscore.default", 1500,
                "starting ELO score"),
        ELO_MAXIMUM(Type.INT, "eloscore.maximum", 3000,
                "max possible ELO score"),
        ELO_KFACTOR(Type.COMMENT, "eloscore.k-factor", null,
                "K-Factor settings"),
        ELO_K_BELOW(Type.INT, "eloscore.k-factor.below", 32,
                "K-Factor below threshold"),
        ELO_K_ABOVE(Type.INT, "eloscore.k-factor.above", 16,
                "K-Factor above threshold"),
        ELO_K_THRESHOLD(Type.INT, "eloscore.k-factor.threshold", 2000,
                "K-Factor threshold"),

        MESSAGES(Type.COMMENT, "msg", null,
                "=== [ Message Settings ] ==="),
        MESSAGES_OVERRIDES(Type.BOOLEAN, "msg.overrides", false, "activate the following overrides"),
        MESSAGES_OVERRIDE_LIST(Type.LIST, "msg.main",
                Arrays.asList("&cName: &7%n", "&cKills: &7%k", "&cDeaths: &7%d", "&cRatio: &7%r", "&cStreak: &7%s", "&cMax Streak: &7%m"), null),

        UPDATE(Type.COMMENT, "update", null,
                "=== [ Updater Settings ] ==="),
        UPDATE_MODE(Type.STRING, "update.mode", "both",
                "what to do? Valid values: disable, announce, download, both"),
        UPDATE_TYPE(Type.STRING, "update.type", "beta",
                "which type of branch to get updates? Valid values: dev, alpha, beta, release"),

        OTHER(Type.COMMENT, "other", null,
                "=== [ Other Features ] ==="),
        OTHER_DISPLAYNAMES(Type.BOOLEAN, "other.displayNames", false, "use players' display names"),
        OTHER_PVPARENA(Type.BOOLEAN, "other.PVPArena", false, "count PVP Arena deaths"),
        OTHER_OP_MESSAGES(Type.BOOLEAN, "other.OPMessages", true, "provide helpful debug messages for new installations"),

        BSTATS(Type.COMMENT, "bStats", null,
                "=== [ bStats Metrics Settings ] ==="),
        BSTATS_ENABLED(Type.BOOLEAN, "bStats.enabled", true,
                "Should we send stats at all? Please keep this in so we have an overview of spread of versions <3"),
        BSTATS_FULL(Type.BOOLEAN, "bStats.full", true,
                "This setting sends a bit more detailed information about which features are used at all. Thank you for supporting me!"),

        IGNORE_WORLDS(Type.LIST, "ignoreworlds", Collections.singletonList("doNotTrack"), "world names where not to count statistics"),

        LEADERBOARDS(Type.MAP, "leaderboards", null, "Leaderboards go here. Do not change these unless you know what you are doing!"),

        VERSION(Type.INT, "ver", null, "Version for automatic config updating"),
        DEBUG(Type.STRING, "debug", null, "Debug - only change this when you know what you are doing"),

        DISPLAYS(Type.LIST, "leaderboards", new ArrayList<String>(), "locations of leaderboards");


        static {
            MYSQL_USERNAME.secret = true;
            MYSQL_PASSWORD.secret = true;
            MYSQL_PORT.secret = true;
            MYSQL_DATABASE.secret = true;
            MYSQL_HOST.secret = true;
        }

        final ConfigEntry.Type type;
        final String node;
        final Object value;
        final String comment;
        public boolean secret;

        Entry(final ConfigEntry.Type type, final String node, final Object def, final String comment) {
            this.type = type;
            this.node = node;
            value = def;
            this.comment = comment;
        }

        /**
         * Try to get a ConfigEntry based on a node string
         *
         * @param node the node to search for
         * @return the entry or null if not found
         */
        public static Entry getByNode(final String node) {
            for (Entry c : values()) {
                if (c.getNode().equals(node)) {
                    return c;
                }
            }
            return null;
        }


        @Override
        public String getComment() {
            return comment;
        }

        @Override
        public String getNode() {
            return node;
        }

        @Override
        public Object getValue() {
            return value;
        }

        @Override
        public Type getType() {
            return type;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }
    }

    /**
     * The Config constructor;
     *
     * Create the config and append comments
     *
     * @param plugin the JavaPlugin which contains the config
     */
    public Config(final CorePlugin plugin) {
        super(plugin, "PVP Stats Config", new File(plugin.getDataFolder(), "config.yml"));

        emptyNodes = new String[]{
                "MySQL", "SQLite", "YML", "statistics", "other",
                "eloscore", "eloscore.k-factor", "msg", "update", "bStats", "general"
        };
    }

    @Override
    public ConfigEntry getByNode(String node) {
        for (final Entry e : Entry.values()) {
            if (e.node.equals(node)) {
                return e;
            }
        }
        return null;
    }

    /**
     * Load the config-file into the YamlConfiguration, and then populate the
     * value maps.
     *
     * @return true, if the load succeeded, false otherwise.
     */
    public boolean load() {
        try {
            cfg.load(configFile);
            reloadMaps();
            return true;
        } catch (final Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    @Override
    protected boolean checkMaterials(String s) {
        // not needed
        return false;
    }

    @Override
    protected void loadMaterials() {
        // not needed
    }
}
